### ISSUE TITLE
replacement gives error in IndexedInduction

### ISSUE THREAD

**Issue by siddhartha-gadgil** [siddhartha-gadgil/ProvingGround#153]
Discovered while importing from Lean. We have
```scala
@ inn 
res85: Term = induc(eq($ajpezd)($ajpezg)($ajpezh)($ajpezh))(($ajpezu : $ajpezd) ↦ ($ajpezv : eq($ajpezd)($ajpezg)($ajpezu)) ↦ $ajpezf($ajpezu))($ajpezj)

@ x 
res86: Term = $ajpezd

@ ar 
res87: Term = $ajpheo

@ val fail3 = Try(inn.replace(x, ar)).failed.get.asInstanceOf[ApplnFailException] 
fail3: ApplnFailException = provingground.HoTT$ApplnFailException: function func  cannot act on given term

@ inn.replace(x, ar) 
provingground.HoTT$ApplnFailException: function func  cannot act on given term
  provingground.HoTT$FuncLike.apply(HoTT.scala:1121)
  provingground.HoTT$FuncLike.apply$(HoTT.scala:1117)
  provingground.HoTT$LambdaTerm.apply(HoTT.scala:1471)
  provingground.induction.TypFamilyMap$FuncTypFamilyMap.typRestrict(TypFamily.scala:536)
  provingground.induction.TypFamilyMap$FuncTypFamilyMap.typRestrict(TypFamily.scala:504)
  provingground.induction.IndexedConstructorPatternMap$IndexedIdMap.inducDataTyp(IndexedConstructorPatternMap.scala:103)
  provingground.induction.IndexedConstructorSeqMap$Cons.inducData(IndexedConstructorSequence.scala:142)
  provingground.induction.IndexedConstructorSeqMap$Cons.indDataCons(IndexedConstructorSequence.scala:148)
  provingground.induction.IndexedConstructorSeqMap$Cons.$anonfun$indDataCons$3(IndexedConstructorSequence.scala:155)
  provingground.induction.IndexedInductiveDefinition$DataCons.subs(InductiveCaseDefinition.scala:202)
  provingground.induction.IndexedInductiveDefinition$Funcs.subs(InductiveCaseDefinition.scala:150)
  provingground.induction.IndexedInductiveDefinition$Funcs.subs(InductiveCaseDefinition.scala:118)
  provingground.HoTT$Subs.replace(HoTT.scala:133)
  provingground.HoTT$Subs.replace$(HoTT.scala:118)
  provingground.induction.IndexedInductiveDefinition$Funcs.replace(InductiveCaseDefinition.scala:118)
  ammonite.$sess.cmd89$.<init>(cmd89.sc:1)
  ammonite.$sess.cmd89$.<clinit>(cmd89.sc)


@ fail3.func 
res90: Term = ($ajpezu : $ajpheo) ↦ ($ajpezv : eq($ajpheo)($ajpezg)($ajpezu)) ↦ $ajpezf($ajpezu)

@ fail3.arg 
res91: Term = $ajpezg
```

**Comment by siddhartha-gadgil** [comment]
A mistake: initialization order in indexed induction

**Comment by siddhartha-gadgil** [comment]
Code to replicate the above error.
```scala
import interface._, LeanInterface._
val mods = getMods("data/group.export")
val env109 = LeanToTerm.fromMods(mods.take(109))
import trepplein._
val d = mods(109).asInstanceOf[DefMod].defn
val fail = env109.parse(d.value, Vector()).failed.get.asInstanceOf[LeanContextException]
val inner = value(value(value(value(value(value(fn))))))
import Fold._
val Vector(fe, ae) = fail.args
for {
          func <- rec(a, vars)
          arg  <- rec(b, vars)
          _ = fn = func
          _ = ar = arg
          res  <- Try(applyFunc(func, arg))
        } yield res
var fn : Term = Type
var ar : Term = Type
for {
          func <- rec(a, vars)
          arg  <- rec(b, vars)
          _ = fn = func
          _ = ar = arg
          res  <- Try(applyFunc(func, arg))
        } yield res
import env109.{parse => rec}
import fail.vars
for {
          func <- rec(a, vars)
          arg  <- rec(b, vars)
          _ = fn = func
          _ = ar = arg
          res  <- Try(applyFunc(func, arg))
        } yield res
val a = fe
val b = ae
for {
          func <- rec(a, vars)
          arg  <- rec(b, vars)
          _ = fn = func
          _ = ar = arg
          res  <- Try(applyFunc(func, arg))
        } yield res
import scala.util.Try
for {
          func <- rec(a, vars)
          arg  <- rec(b, vars)
          _ = fn = func
          _ = ar = arg
          res  <- Try(applyFunc(func, arg))
        } yield res
val inner = value(value(value(value(value(value(fn))))))
val inn = FormalAppln.unapply(inner).get._1
val x = variable(fn)
inn.replace(x, ar)
```

**Comment by siddhartha-gadgil** [comment]
Error narrowed down to `replacement` method for `DataCons` in the indexed inductive case.

**Comment by siddhartha-gadgil** [comment]
Looks like a case of substitution not deep enough. Continuing the above:
```scala
@ val fail = Try(inn.replace(x, ar)).failed.get.asInstanceOf[ApplnFailException]
   
fail: ApplnFailException = provingground.HoTT$ApplnFailException: function func  cannot act on given term

@ fail.func 
res58: Term = ($jxpedy : $jxpfnb) ↦ ($jxpedz : eq($jxpfnb)($jxpedk)($jxpedy)) ↦ $jxpedj($jxpedy)

@ fail.arg 
res59: Term = $jxpedk

@ fail.arg.typ 
res60: Typ[U] = $jxpedh
```

**Comment by siddhartha-gadgil** [comment]
fixed in e9d43fb22ff
